File: packages/wasmvm/wasmlib/src/wasmtypes/scbytes.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use std::convert::TryInto;

use crate::*;

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub fn bytes_decode(dec: &mut WasmDecoder) -> Vec<u8> {
    dec.bytes().to_vec()
}

pub fn bytes_encode(enc: &mut WasmEncoder, value: &[u8])  {
    enc.bytes(value);
}

pub fn bytes_from_bytes(buf: &[u8]) -> Vec<u8> {
    buf.to_vec()
}

pub fn bytes_to_bytes(value: &[u8]) -> Vec<u8> {
    value.to_vec()
}

pub fn bytes_to_string(value: &[u8]) -> String {
    // TODO standardize human readable string
    base58_encode(value)
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub struct ScImmutableBytes {
    proxy: Proxy
}

impl ScImmutableBytes {
    pub fn new(proxy: Proxy) -> ScImmutableBytes {
        ScImmutableBytes { proxy }
    }

    pub fn exists(&self) -> bool {
        self.proxy.exists()
    }

    pub fn to_string(&self) -> String {
        bytes_to_string(&self.value())
    }

    pub fn value(&self) -> Vec<u8> {
        bytes_from_bytes(&self.proxy.get())
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

// value proxy for mutable ScBytes in host container
pub struct ScMutableBytes {
    proxy: Proxy
}

impl ScMutableBytes {
    pub fn new(proxy: Proxy) -> ScMutableBytes {
        ScMutableBytes { proxy }
    }

    pub fn delete(&self)  {
        self.proxy.delete();
    }

    pub fn exists(&self) -> bool {
        self.proxy.exists()
    }

    pub fn set_value(&self, value: &[u8]) {
        self.proxy.set(&bytes_to_bytes(value));
    }

    pub fn to_string(&self) -> String {
        bytes_to_string(&self.value())
    }

    pub fn value(&self) -> Vec<u8> {
        bytes_from_bytes(&self.proxy.get())
    }
}

File: packages/wasmvm/wasmlib/as/wasmlib/wasmtypes/scbytes.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {hexDecode, hexEncode, WasmDecoder, WasmEncoder} from "./codec";
import {Proxy} from "./proxy";

export function bytesCompare(lhs: Uint8Array, rhs: Uint8Array): i32 {
    const size = (lhs.length < rhs.length) ? lhs.length : rhs.length;
    for (let i = 0; i < size; i++) {
        if (lhs[i] != rhs[i]) {
            return (lhs[i] > rhs[i]) ? 1 : -1;
        }
    }
    return (lhs.length > size) ? 1 : (rhs.length > size) ? -1 : 0;
}

export function bytesEquals(lhs: Uint8Array, rhs: Uint8Array): bool {
    if (lhs.length != rhs.length) {
        return false;
    }
    for (let i = 0; i < lhs.length; i++) {
        if (lhs[i] != rhs[i]) {
            return false;
        }
    }
    return true;
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export function bytesDecode(dec: WasmDecoder): Uint8Array {
    return dec.bytes();
}

export function bytesEncode(enc: WasmEncoder, value: Uint8Array): void {
    enc.bytes(value);
}

export function bytesFromBytes(buf: Uint8Array): Uint8Array {
    return buf;
}

export function bytesFromUint8Array(arr: Uint8Array): Uint8Array {
    return arr.slice();
}

export function bytesToBytes(buf: Uint8Array): Uint8Array {
    return buf;
}

export function bytesToUint8Array(buf: Uint8Array): Uint8Array {
    return buf.slice();
}

export function bytesFromString(value: string): Uint8Array {
    return hexDecode(value);
}

export function bytesToString(value: Uint8Array): string {
    return hexEncode(value);
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScImmutableBytes {
    proxy: Proxy;

    constructor(proxy: Proxy) {
        this.proxy = proxy;
    }

    exists(): bool {
        return this.proxy.exists();
    }

    toString(): string {
        return bytesToString(this.value());
    }

    value(): Uint8Array {
        return bytesFromBytes(this.proxy.get());
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScMutableBytes extends ScImmutableBytes {
    delete(): void {
        this.proxy.delete();
    }

    setValue(value: Uint8Array): void {
        this.proxy.set(bytesToBytes(value));
    }
}
